File: src/python/dab/state.py
# dab/state.py — state and initial_state
from dataclasses import dataclass
from typing import FrozenSet

from dab.types import Edge


@dataclass(frozen=True)
class State:
    nx: int
    ny: int
    edges: FrozenSet[Edge]
    s1: int
    s2: int
    player: int  # 1 or 2


def initial_state(nx: int, ny: int) -> State:
    return State(nx=nx, ny=ny, edges=frozenset(), s1=0, s2=0, player=1)

File: src/java/src/main/java/dab/State.java
package dab;

import java.util.HashSet;
import java.util.Set;

public record State(int nx, int ny, Set<Edge> edges, int s1, int s2, int player) {
    public static State initial(int nx, int ny) {
        if (nx < 1 || ny < 1) throw new IllegalArgumentException("nx, ny >= 1");
        return new State(nx, ny, new HashSet<>(), 0, 0, 1);
    }

    public State withMoveApplied(Edge e, int addS1, int addS2, int nextPlayer) {
        Set<Edge> next = new HashSet<>(edges);
        next.add(Edge.normalized(e.a(), e.b()));
        return new State(nx, ny, next, s1 + addS1, s2 + addS2, nextPlayer);
    }
}
